File: lib/views/sensor_details_page.dart
import 'package:flutter/material.dart';

/// A page that displays details for a specific sensor and allows
/// users to add data points related to energy usage.
///
/// This page provides an interface for viewing existing data points
/// and adding new ones, along with options to delete the sensor.
class SensorDetailsPage extends StatefulWidget {
  final Map<String, dynamic> sensorData; // Sensor data passed from the previous page
  final Function(Map<String, dynamic>) onUpdate; // Callback to update the sensor data
  final Function onDelete; // Callback to delete the sensor

  SensorDetailsPage({
    required this.sensorData,
    required this.onUpdate,
    required this.onDelete,
  });

  @override
  _SensorDetailsPageState createState() => _SensorDetailsPageState();
}

class _SensorDetailsPageState extends State<SensorDetailsPage> {
  late Map<String, dynamic> sensorData; // Local copy of sensor data
  final TextEditingController _energyUsageController = TextEditingController(); // Controller for energy usage input
  final TextEditingController _additionalFieldController = TextEditingController(); // Controller for additional fields
  final TextEditingController _timestampController = TextEditingController(); // Controller for timestamp input

  @override
  void initState() {
    super.initState();
    sensorData = widget.sensorData; // Initialize sensor data
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Sensor Details: ${sensorData['sensor_id']}'), // Display sensor ID in the app bar
        actions: [
          // Button to delete the sensor
          IconButton(
            icon: Icon(Icons.delete),
            onPressed: () {
              widget.onDelete(); // Call the delete callback
              Navigator.pop(context); // Navigate back to the previous page
            },
          ),
        ],
      ),
      body: Column(
        children: [
          // Display sensor data points in a list
          Expanded(
            child: ListView.builder(
              itemCount: sensorData['data_points'].length,
              itemBuilder: (context, index) {
                final dataPoint = sensorData['data_points'][index]; // Get data point
                return ListTile(
                  title: Text('Timestamp: ${dataPoint['timestamp']}'), // Display timestamp
                  subtitle: Text('Energy Usage: ${dataPoint['energy_usage']} kWh'), // Display energy usage
                );
              },
            ),
          ),
          Divider(),
          // Form to add a new data point
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: Column(
              children: [
                // Input field for Timestamp
                TextField(
                  controller: _timestampController,
                  decoration: InputDecoration(labelText: 'Timestamp (Unix Epoch)'),
                  keyboardType: TextInputType.number,
                  onTap: _pickDateTime, // Open date/time picker
                ),
                // Input field for Energy Usage
                TextField(
                  controller: _energyUsageController,
                  decoration: InputDecoration(labelText: 'Energy Usage (kWh)'),
                  keyboardType: TextInputType.number,
                ),
                // Additional fields based on sensor category
                if (sensorData['category'] == 'HVAC' || sensorData['category'] == 'Outlets')
                  TextField(
                    controller: _additionalFieldController,
                    decoration: InputDecoration(labelText: 'Temperature (Celsius)'),
                    keyboardType: TextInputType.number,
                  ),
                if (sensorData['category'] == 'Lighting')
                  TextField(
                    controller: _additionalFieldController,
                    decoration: InputDecoration(labelText: 'Lighting Level (Lux)'),
                    keyboardType: TextInputType.number,
                  ),
                if (sensorData['category'] == 'Hot Water')
                  TextField(
                    controller: _additionalFieldController,
                    decoration: InputDecoration(labelText: 'Water Usage (Liters)'),
                    keyboardType: TextInputType.number,
                  ),
                // Button to add the new data point
                ElevatedButton(
                  onPressed: _addDataPoint,
                  child: Text('Add Data Point'),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  /// Opens a date and time picker to set the timestamp for the data point.
  void _pickDateTime() async {
    // Implementation for picking a date and time
  }

  /// Adds a new data point to the sensor's data and updates the parent widget.
  void _addDataPoint() {
    final dataPoint = {
      'timestamp': _timestampController.text,
      'energy_usage': _energyUsageController.text,
    };

    // Add additional fields based on the sensor category
    if (sensorData['category'] == 'HVAC' || sensorData['category'] == 'Outlets') {
      dataPoint['temperature'] = _additionalFieldController.text;
    } else if (sensorData['category'] == 'Lighting') {
      dataPoint['lighting_level'] = _additionalFieldController.text;
    } else if (sensorData['category'] == 'Hot Water') {
      dataPoint['water_usage'] = _additionalFieldController.text;
    }

    setState(() {
      sensorData['data_points'].add(dataPoint); // Add the new data point
    });

    // Update the sensor data in the parent widget
    widget.onUpdate(sensorData);

    // Clear input fields after adding the data point
    _timestampController.clear();
    _energyUsageController.clear();
    _additionalFieldController.clear();
  }
}

File: frontend/src/components/SensorDetailsPage.js
import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';

function SensorDetailsPage({ sensorData: initialPropData, onUpdate, onDelete}) {
    // Retrieve sensor data and callbacks from location state if not provided as props
    const location = useLocation();
    const initialSensorData = initialPropData || (location.state && location.state.sensorData);


    // Local copy of sensor data
    const [sensorData, setSensorData] = useState(initialSensorData || {});
    // Input states for the new data point
    const [timestamp, setTimestamp] = useState('');
    const [energyUsage, setEnergyUsage] = useState('');
    const [additionalField, setAdditionalField] = useState('');


    // If no sensor data is provided, render a fallback message.
    if (!initialSensorData) {
        return <div>No sensor data provided.</div>;
    }

    // Stub for picking a date/time (you can integrate a datetime picker library here)
    const pickDateTime = () => {
        // For example, you might open a modal or use an HTML date input.
    };

    // Adds a new data point to the sensor's data
    const addDataPoint = () => {
        const dataPoint = {
            timestamp: timestamp,
            energy_usage: energyUsage,
        };

        // Add additional fields based on the sensor category
        if (sensorData.category === 'HVAC' || sensorData.category === 'Outlets') {
            dataPoint.temperature = additionalField;
        } else if (sensorData.category === 'Lighting') {
            dataPoint.lighting_level = additionalField;
        } else if (sensorData.category === 'Hot Water') {
            dataPoint.water_usage = additionalField;
        }

        // Create an updated sensor object with the new data point added.
        const updatedSensorData = {
            ...sensorData,
            dataPoints: [...sensorData.dataPoints, dataPoint],
        };

        setSensorData(updatedSensorData);
        if (onUpdate) {
            onUpdate(updatedSensorData);
        }

        // Clear the input fields
        setTimestamp('');
        setEnergyUsage('');
        setAdditionalField('');
    };

    // Handle sensor deletion and navigate back
    const handleDelete = () => {
        if (onDelete) {
            onDelete();
        }
        window.history.back();
    };

    return (
        <div style={{ padding: '16px', fontFamily: 'sans-serif' }}>
            {/* App Bar */}
            <header style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
                <h1>Sensor Details: {sensorData.sensor_id}</h1>
                <button onClick={handleDelete}>Delete</button>
            </header>

            {/* Data Points List */}
            <div style={{ maxHeight: '300px', overflowY: 'auto', margin: '16px 0' }}>
                <ul>
                    {sensorData.dataPoints.map((dataPoint, index) => (
                        <li key={index} style={{ marginBottom: '8px' }}>
                            <div><strong>Timestamp:</strong> {dataPoint.timestamp}</div>
                            <div><strong>Energy Usage:</strong> {dataPoint.energy_usage} kWh</div>
                        </li>
                    ))}
                </ul>
            </div>

            <hr />

            {/* Form to add a new data point */}
            <div style={{ marginTop: '16px' }}>
                <div style={{ marginBottom: '8px' }}>
                    <label>
                        Timestamp (Unix Epoch):{' '}
                        <input
                            type="number"
                            value={timestamp}
                            onChange={(e) => setTimestamp(e.target.value)}
                            onFocus={pickDateTime}
                        />
                    </label>
                </div>
                <div style={{ marginBottom: '8px' }}>
                    <label>
                        Energy Usage (kWh):{' '}
                        <input
                            type="number"
                            value={energyUsage}
                            onChange={(e) => setEnergyUsage(e.target.value)}
                        />
                    </label>
                </div>

                {/* Conditionally render additional field based on sensor category */}
                {(sensorData.category === 'HVAC' || sensorData.category === 'Outlets') && (
                    <div style={{ marginBottom: '8px' }}>
                        <label>
                            Temperature (Celsius):{' '}
                            <input
                                type="number"
                                value={additionalField}
                                onChange={(e) => setAdditionalField(e.target.value)}
                            />
                        </label>
                    </div>
                )}
                {sensorData.category === 'Lighting' && (
                    <div style={{ marginBottom: '8px' }}>
                        <label>
                            Lighting Level (Lux):{' '}
                            <input
                                type="number"
                                value={additionalField}
                                onChange={(e) => setAdditionalField(e.target.value)}
                            />
                        </label>
                    </div>
                )}
                {sensorData.category === 'Hot Water' && (
                    <div style={{ marginBottom: '8px' }}>
                        <label>
                            Water Usage (Liters):{' '}
                            <input
                                type="number"
                                value={additionalField}
                                onChange={(e) => setAdditionalField(e.target.value)}
                            />
                        </label>
                    </div>
                )}

                <button onClick={addDataPoint}>Add Data Point</button>
            </div>
        </div>
    );
}

export default SensorDetailsPage;
